src/e3/testsuite/control.py

```diff
@@ -0,0 +1,251 @@
+
+    def default_opt_results(self, driver):
+        """
+        Return the default options. test.opt files can override these.
+
+        By default, a test is not DEAD, SKIP, nor XFAIL. Its execution timeout
+        is 780 seconds. Test script is "test.cmd" and its output is compared
+        against "test.out".
+        """
+        return {
+            "RLIMIT": "780",
+            "DEAD": None,
+            "XFAIL": None,
+            "SKIP": None,
+            "OUT": "test.out",
+            "CMD": self.default_script(driver),
+            "FILESIZE_LIMIT": None,
+            "TIMING": None,
+            "NOTE": None,
+        }
+
+    def __init__(self, system_tags, opt_filename="test.opt"):
+        """
+        :param str|list[str] system_tags: Tags to forward to OptFileParse().
+        :param str opt_filename: Name of the file to parse, relative to the
+            test directory.
+        """
+        self.opt_filename = opt_filename
+        self.system_tags = system_tags
+
+    def create(self, driver):
+        # If there is a "control" entry in the testcase's test.yaml file while
+        # the optfile mechanism is in use, it probably means someone mistakenly
+        # wrote a "control" entry that will not be interpreted: be helpful and
+        # warn about it.
+        if "control" in driver.test_env:
+            logger.warning(
+                '{}: "control" entry found in test.yaml whereas only test.opt'
+                ' files are considered'.format(driver.test_env["test_name"])
+            )
+
+        # If it exists, parse the "test.opt" file in the test directory.
+        # Create a dummy optfile otherwise.
+        filename = driver.test_dir(self.opt_filename)
+        optfile = (OptFileParse(self.system_tags, filename)
+                   if os.path.exists(filename)
+                   else OptFileParse(self.system_tags, []))
+
+        # Create a TestControl depending on the contents of the optfile
+        message = None
+        skip = False
+        xfail = False
+        opt_results = optfile.get_values(self.default_opt_results(driver))
+        if opt_results["DEAD"] is not None:
+            message = opt_results["DEAD"]
+            skip = True
+        elif opt_results["SKIP"] is not None:
+            message = opt_results["SKIP"]
+            skip = True
+            xfail = True
+        elif opt_results["XFAIL"] is not None:
+            message = opt_results["XFAIL"]
+            xfail = True
+        result = TestControl(message, skip, xfail)
+
+        # Store results in the TestControl instance, so that "driver" has
+        # access to it.
+        result.opt_results = opt_results
+
+        return result
+
+
+class AdaCoreLegacyTestControlCreator(OptfileTestControlCreator):
+    """
+    Create test controls for legacy AdaCore testsuites, for "test.opt" files.
+    """
+
+    def default_script(self, driver):
+        # Return the filename for the first script file that exists. The lookup
```

Comment:
use docstring

Reply:
No need: this is an overriding method which does exactly what the base method documents